src/coinsselectionalgorithm.cpp

```diff
@@ -0,0 +1,596 @@
+    optimalSelection = std::vector<char>(problemDimension, 0);
+
+    optimalTotalAmount = 0;
+    optimalTotalSize = 0;
+    optimalTotalSelection= 0;
+
+    isRunning = false;
+    stopRequested = false;
+    solvingThread = nullptr;
+    hasCompleted = false;
+    #if COINS_SELECTION_ALGORITHM_PROFILING
+    executionMicroseconds = 0;
+    #endif
+}
+
+CCoinsSelectionAlgorithmBase::~CCoinsSelectionAlgorithmBase()
+{
+    StopSolving();
+}
+
+std::vector<CAmount> CCoinsSelectionAlgorithmBase::PrepareAmounts(std::vector<std::pair<CAmount, size_t>>& amountsAndSizes)
+{
+    std::sort(amountsAndSizes.begin(), amountsAndSizes.end(), std::greater<>());
+    std::vector<CAmount> sortedAmounts(amountsAndSizes.size(), 0);
+    for (int index = 0; index < problemDimension; ++index)
+    {
+        sortedAmounts[index] = amountsAndSizes[index].first;
+    }
+    return sortedAmounts;
+}
+
+std::vector<size_t> CCoinsSelectionAlgorithmBase::PrepareSizes(std::vector<std::pair<CAmount, size_t>>& amountsAndSizes)
+{
+    std::vector<size_t> sortedSizes(amountsAndSizes.size(), 0);
+    for (int index = 0; index < problemDimension; ++index)
+    {
+        sortedSizes[index] = amountsAndSizes[index].second;
+    }
+    return sortedSizes;
+}
+
+bool CCoinsSelectionAlgorithmBase::Reset()
+{
+    bool resetActuallyDone = false;
+
+    if (isRunning || hasCompleted)
+    {
+        if (isRunning)
+        {
+            StopSolving();
+        }
+        for (int index = 0; index < problemDimension; ++index)
+        {
+            tempSelection[index] = 0;
+            optimalSelection[index] = 0;
+        }
+
+        optimalTotalAmount = 0;
+        optimalTotalSize = 0;
+        optimalTotalSelection= 0;
+
+        isRunning = false;
+        stopRequested = false;
+        solvingThread = nullptr;
+        hasCompleted = false;
+        #if COINS_SELECTION_ALGORITHM_PROFILING
+        executionMicroseconds = 0;
+        #endif
+        resetActuallyDone = true;
+    }
+
+    return resetActuallyDone;
+}
+
+std::string CCoinsSelectionAlgorithmBase::ToString()
+{
+    return std::string("Input:")
+                                + "{targetAmount=" + std::to_string(targetAmount) +
+                                  ",targetAmountPlusOffset=" + std::to_string(targetAmountPlusOffset) +
+                                  ",availableTotalSize=" + std::to_string(availableTotalSize) + "}\n" +
+                       "Output:"
+                                + "{optimalTotalAmount=" + std::to_string(optimalTotalAmount) +
+                                  ",optimalTotalSize=" + std::to_string(optimalTotalSize) +
+                                  ",optimalTotalSelection=" + std::to_string(optimalTotalSelection) + "}\n";
+}
+
+void CCoinsSelectionAlgorithmBase::StartSolvingAsync()
+{
+    if (!isRunning)
+    {
+        solvingThread = std::unique_ptr<std::thread>(new std::thread(&CCoinsSelectionAlgorithmBase::Solve, this));
+    }
```

Comment:
isRunning MUST be set before starting a new thread.
If not this may happen :
- StartSolvingAsync is invoked when isRunning == false
- The scheduler takes some time to prepare the thread and synchronize the memory: also Solve() sets the atomic var to true AFTER having done Reset() so it takes time
- In the mean time another StartSolvingAsync is invoked : isRunning is still false and solvingThread is reset to a new thread making the original thread orphan.

Reply:
In the context the var `isRunning` is totally useless: is enough to check whether solvingThread has a valid pointer to a thread

Reply:
I agree with the criticality described in your scenario.
I cannot get rid of the variable `isRunning` because it is used to check if the sync solving is in progress; this means that checking for `solvingThread` is pointless if a sync solution has been requested through a direct call to `Solve` method. I have added another control variable `asyncStartRequested` that should handle properly the case you highlighted.